ruby/ql/src/queries/security/cwe-020/IncompleteHostnameRegExp.ql

```diff
@@ -0,0 +1,55 @@
+/**
+ * @name Incomplete regular expression for hostnames
+ * @description Matching a URL or hostname against a regular expression that contains an unescaped dot as part of the hostname might match more hostnames than expected.
+ * @kind problem
+ * @problem.severity warning
+ * @security-severity 7.8
+ * @precision high
+ * @id rb/incomplete-hostname-regexp
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-20
+ */
+
+private import ruby
+private import codeql.ruby.ApiGraphs
+
+private string commonTopLevelDomainRegex() { result = "com|org|edu|gov|uk|net|io" }
+
+/**
+ * Holds if `pattern` is a regular expression pattern for URLs with a host matched by `hostPart`,
+ * and `pattern` contains a subtle mistake that allows it to match unexpected hosts.
+ */
+bindingset[pattern]
+predicate isIncompleteHostNameRegExpPattern(string pattern, string hostPart) {
+  hostPart =
+    pattern
+        .regexpCapture("(?i).*" +
+            // an unescaped single `.`
+            "(?<!\\\\)[.]" +
+            // immediately followed by a sequence of subdomains, perhaps with some regex characters mixed in, followed by a known TLD
+            "([():|?a-z0-9-]+(\\\\)?[.](" + commonTopLevelDomainRegex() + "))" + ".*", 1)
+}
```

Comment:
I wonder, is it worth adding a shared `HostnameRegExpUtils.qll` file to share this with the python implementation? I notice the [JS version](https://github.com/aibaars/codeql/blob/main/javascript/ql/src/Security/CWE-020/HostnameRegexpShared.qll) has a different implementation. I don't know whether we (the dynamic teams) want to consider consolidating them?

Reply:
Good idea.

Reply:
A bit of history: The one you copied originated in JS, written before we parsed string literals as RegExps. When we [started extracting strings](https://github.com/github/codeql/pull/2252) as RegExps, it was rewritten to its current form.

I'd say the new version is better, if you have ASTs for regexps.